docs/howtos/configureportrange.md
---
layout: default
title: Configure Port range
parent: How to's
nav_order: 7
---

# Configure Thundernetes port range

By default, Thundernetes will allocate ports in the range 10000-12000 to your GameServers. These ports are allocated to the entire set of VMs in the cluster and are open for each and every VM. If you need more or just a different port range, you can configure it via changing the `MIN_PORT` and the `MAX_PORT` environment variables in the controller deployment YAML file. However, do not modify the port range when there game servers running on the cluster, since this will probably corrupt the port registry, especially if the new and the old range are different.

CHANGELOG.md
# Changelog

## 0.2.0 / 2022-01-13

* [FEATURE] Updated README.md by @AmieDD in https://github.com/PlayFab/thundernetes/pull/64
* [ENHANCEMENT] Update GameServer.Status.State when initializing by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/68
* [ENHANCEMENT] Adding a CHANGELOG to track updates to the project. by @khaines in https://github.com/PlayFab/thundernetes/pull/90
* [ENHANCEMENT] adding node internal ip + node name by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/92
* [ENHANCEMENT] Create CODE_OF_CONDUCT.md by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/94
* [FEATURE] GameServer API by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/91
* [FEATURE] adding connected player names by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/89
* [BUGFIX] fixing GameServerBuild controller issues by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/95
* [ENHANCEMENT] Initial Contributors documentation by @khaines in https://github.com/PlayFab/thundernetes/pull/96
* [ENHANCEMENT] replacing update with patch on pod annotations by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/103
* [FEATURE] Sidecar to Daemonset by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/97
* [FEATURE] Enable metrics endpoint on node agent by @khaines in https://github.com/PlayFab/thundernetes/pull/111
* [ENHANCEMENT] .net 6 for .NET sample and openarena by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/112
* [BUGFIX] fixing bug on counters by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/115
* [BUGFIX] fixes for Unity GSDK by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/117
* [ENHANCEMENT] Reorganizing Project Layout by @khaines in https://github.com/PlayFab/thundernetes/pull/119
* [FEATURE] Support setting GameServers Pods to host network by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/122
* [ENHANCEMENT] schedule NodeAgent on GameServer Pods by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/124
* [ENHANCEMENT] adding information about Pod Priority by @dgkanatsios in https://github.com/PlayFab/thundernetes/pull/125

## 0.1.0 / 2021-11-12

* [CHANGE] remove deprecated .NET Core sidecar. #2
* [CHANGE] renaming CRD instance to CR. #62
* [FEATURE] Adding Cluster Autoscaler annotations to game server pods to prevent eviction of active sessions. #20
* [FEATURE] adding instructions for running end to end tests on macOS. #19
* [FEATURE] adding some architecture notes. #21
* [FEATURE] namespace stuck on terminating state - FAQ. #29
* [FEATURE] pod affinity - faq. #30
* [FEATURE] Adding a simple example dashboard for Grafana. #33
* [FEATURE] adding information on fluentbit. #34
* [FEATURE] adding guidance for an additional Node Pool. #24
* [FEATURE] allocator draft. #28
* [FEATURE] switching to watches for allocation. #31
* [FEATURE] adding NodeAffinity sample. #39
* [FEATURE] Created prerequisites.md. #51
* [FEATURE] Better tracking ConnectedPlayers. #48
* [FEATURE] adding logrus to initcontainer. #50
* [FEATURE] adding uninstall information. #55
* [FEATURE] Created troubleshooting guide and categories. #63
* [ENHANCEMENT] Update yourgameserver.md. #23
* [ENCHANTMENT] adding resources for samples. #15
* [ENHANCEMENT] Consolidation of Gauge metrics for simplicity. #41
* [ENHANCEMENT] Updated quickstart.md. #44
* [ENHANCEMENT] Monitoring & Grafana Updates. #42
* [ENHANCEMENT] updating OpenArena sample readme. #3
* [BUGFIX] fixing PodAffinity FAQ. #59
* [BUGFIX] fix for GsdkConfig. #58


cmd/standby-forecaster/config.go
package main

import (
	"flag"
	"path/filepath"
	"time"

	"k8s.io/client-go/util/homedir"
)

const (
	defaultMetricQuery = "sum(simulated_instances)"
)

type Config struct {
	QueryUrl                               string        `yaml:"queryUrl"`
	MetricQuery                            string        `yaml:"metricQuery"`
	ForecastFrequency                      time.Duration `yaml:"frequency"`
	ForecastedPoints                       int           `yaml:"forecastedPoints"`
	Port                                   int           `yaml:"port"`
	HistoricalQueryRange                   time.Duration `yaml:"historicalQueryRange"`
	AlphaConstant                          float64       `yaml:"alpha"`
	BetaConstant                           float64       `yaml:"beta"`
	GammaConstant                          float64       `yaml:"gamma"`
	SeasonLength                           int           `yaml:"seasonLength"`
	EnableHoltWintersForecast              bool          `yaml:"enableHoltWintersForecast"`
	EnableLongLinearForecast               bool          `yaml:"enableLongLinearForecast"`
	EnableShortlinearForecast              bool          `yaml:"enableShortLinearForecast"`
	LongLinearHistoryPoints                int           `yaml:"longLinearHistoryPoints"`
	LongLinearForecastedPoints             int           `yaml:"longLinearForecastedPoints"`
	ShortLinearHistoryPoints               int           `yaml:"shortLinearHistoryPoints"`
	ShortLinearForecastedPoints            int           `yaml:"shortLinearForecastedPoints"`
	ForecastMinimumServers                 int           `yaml:"forecastMinimumServers"`
	ForecastBufferServers                  int           `yaml:"forecastBufferServers"`
	DatapointGranularity                   time.Duration `yaml:"datapointGranularity"`
	EnableScaleDownCircuitBreaker          bool          `yaml:"enableScaleDownCircuitBreaker"`
	MaxScaleDownStepSize                   int           `yaml:"maxScaleDownStepSize"`
	EnableScaleDownCoolOff                 bool          `yaml:"enableScaleDownCoolOff"`
	ScaleDownCoolOffPeriod                 time.Duration `yaml:"scaleDownCoolOffPeriod"`
	TargetGameServerBuildName              string        `yaml:"targetGameServerBuildName"`
	TargetGameServerBuildNamespace         string        `yaml:"targetGameServerBuildNameSpace"`
	MetricToServerConversionOperation      string        `yaml:"metricToServerConversionOperation"`
	MetricToServerConversionOperationValue float64       `yaml:"metricToServerConversionOperationValue"`
	K8s                                    struct {
		RunInCluster bool   `yaml:"runInCluster"`
		KubeConfig   string `yaml:"kubeConfig"`
	} `yaml:"k8s"`
}

func (c *Config) RegisterFlags(f *flag.FlagSet) {
	f.StringVar(&c.QueryUrl, "queryUrl", "http://localhost:9090", "URL to query the Prometheus server")
	f.StringVar(&c.MetricQuery, "metricQuery", defaultMetricQuery, "Query to retrieve the metric")
	f.IntVar(&c.ForecastedPoints, "points", 10, "number of points to forecast")
	f.IntVar(&c.Port, "port", 8080, "port to serve metrics endpoint")
	f.DurationVar(&c.ForecastFrequency, "forecast-frequency", 15*time.Second, "how often to calculate a new forecast")
	f.DurationVar(&c.HistoricalQueryRange, "historical-query-range", 6*24*time.Hour, "historical range for data points")
	f.Float64Var(&c.AlphaConstant, "alpha", 0.6, "alpha constant")
	f.Float64Var(&c.BetaConstant, "beta", 0.6, "beta constant")
	f.Float64Var(&c.GammaConstant, "gamma", 0.45, "gamma constant")
	f.IntVar(&c.SeasonLength, "seasonLength", 60, "season length in data points")
	f.BoolVar(&c.EnableHoltWintersForecast, "enableHoltWintersForecast", true, "enable Holt Winters forecast")
	f.BoolVar(&c.EnableLongLinearForecast, "enableLongLinearForecast", true, "enable Long Linear forecast")
	f.BoolVar(&c.EnableShortlinearForecast, "enableShortLinearForecast", true, "enable Short Linear forecast")
	f.IntVar(&c.LongLinearHistoryPoints, "longLinearHistoryPoints", 30, "number of historical points to use for Long Linear forecast")
	f.IntVar(&c.LongLinearForecastedPoints, "longLinearForecastedPoints", 10, "number of forecasted points to use for Long Linear forecast")
	f.IntVar(&c.ShortLinearHistoryPoints, "shortLinearHistoryPoints", 10, "number of historical points to use for Short Linear forecast")
	f.IntVar(&c.ShortLinearForecastedPoints, "shortLinearForecastedPoints", 5, "number of forecasted points to use for Short Linear forecast")
	f.IntVar(&c.ForecastMinimumServers, "forecastMinimumServers", 1, "minimum number of Standby servers to forecast")
	f.IntVar(&c.ForecastBufferServers, "forecastBufferServers", 0, "number of 'buffer' standby servers to add to forecast")
	f.DurationVar(&c.DatapointGranularity, "datapointGranularity", 1*time.Minute, "granularity of datapoints gathered from Prometheus")
	f.BoolVar(&c.EnableScaleDownCircuitBreaker, "enableScaleDownCircuitBreaker", true, "enable scale down circuit breaker")
	f.IntVar(&c.MaxScaleDownStepSize, "maxScaleDownStepSize", 1, "maximum step size for scale down")
	f.BoolVar(&c.EnableScaleDownCoolOff, "enableScaleDownCoolOff", true, "enable scale down cool off")
	f.DurationVar(&c.ScaleDownCoolOffPeriod, "scaleDownCoolOffPeriod", 5*time.Minute, "scale down cool off period")
	f.StringVar(&c.TargetGameServerBuildName, "targetGameServerBuildName", "gameserverbuild-sample-netcore", "target game server build name")
	f.StringVar(&c.TargetGameServerBuildNamespace, "targetGameServerBuildNamespace", "default", "target game server build namespace")
	f.StringVar(&c.MetricToServerConversionOperation, "metricToServerConversionOperation", "divide", "metric to server conversion operation. Can be 'divide' or 'multiply'")
	f.Float64Var(&c.MetricToServerConversionOperationValue, "metricToServerConversionOperationValue", 10, "metric to server conversion operation value")
	f.BoolVar(&c.K8s.RunInCluster, "k8s.runInCluster", false, "run in cluster")
	defaultKubConfig := ""
	if home := homedir.HomeDir(); home != "" {
		defaultKubConfig = filepath.Join(home, ".kube", "config")
	}
	f.StringVar(&c.K8s.KubeConfig, "k8s.kubeConfig", defaultKubConfig, "(optional) absolute path to the kubeconfig file")
}
